util/design/lib/OtpMemImg.py

```diff
@@ -413,6 +421,52 @@ def streamout_partition(self, part):
         # in the memory hex file.
         return data, annotation
 
+    def validate_data_perm(self, data_perm):
+        '''Validate data permutation option'''
+
+        # Byte aligned total width after adding ECC bits
+        total_bytelen = (
+            (self.lc_state.config['secded']['data_width'] +
+             self.lc_state.config['secded']['ecc_width'] + 7) // 8)
+        total_bitlen = total_bytelen * 8
+
+        # Default mapping
+        if not data_perm:
+            self.data_perm = list(range(total_bitlen))
+
+        # Parse custom mapping
+        else:
+            # Check the format first
+            pattern = r'^((?:\[[0-9]+:[0-9]+\])+(?:,\[[0-9]+:[0-9]+\])*)'
+            match = re.fullmatch(pattern, data_perm)
+            if match is None:
+                raise RuntimeError('Data permutation "{}" has invalid format'
+                                   .format(data_perm))
+            # Expand the ranges
+            self.data_perm = []
+            groups = match.groups()
+            for group in groups[0].split(','):
+                k1, k0 = [int(x) for x in group[1:-1].split(':')]
+                if k1 > k0:
+                    self.data_perm = list(range(k0, k1 + 1)) + self.data_perm
+                else:
+                    self.data_perm = list(range(k0, k1 - 1, -1)) + self.data_perm
```

Comment:
It would probably be slightly neater to put this parsing in `gen-otp-image.py` and use it as a [type converter](https://docs.python.org/3/library/argparse.html#type). Then argparse catches the ValueError that you would throw on failure, giving slightly nicer error messages. But it's not a big deal.

Reply:
Agreed. I've split this conversion part out into a type converter.